### PR TITLE
Feature/checkpoint interval

### PR DESCRIPTION
**Pull request by mingruimingrui**
## Feature Checkpoint Interval

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Now you can save embeddings at regular intervals!

## How Has This Been Tested (if it applies)

No unit tests are created but functionality tested locally to work.

## Checklist

- [X] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [X] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by lerks** [comment]
When I first read the description I thought that what you wanted was to save checkpoints _more_ often, but from the code it appears that you actually want to preserve 1 out of every N checkpoints (where N is the parameter you are adding). I have nothing against the latter, but I'm curious to know what's the reason.

**Comment by mingruimingrui** [comment]
For models with long training durations, having checkpoints allow one to visualise if meaningful embeddings are being learnt during training.

I know that this is the purpose of evaluation sets but high quality evaluation sets are difficult to obtain and often we just sample a subset of all edges and call it the evaluation set. 

For building demos, it can used to display visualisations of how embeddings are learnt which the non-technical personals would often ask to see but you'd be able to ask for more servers this way!

They're not really 'real' reasons. Just a quality of life feature for debugging and misc purposes.

**Comment by adamlerer** [comment]
I think this is reasonable functionality to add, @mingruimingrui is right that you often want to plot learning curves on some test data that doesn't fit nicely into the "eval-during-training" code. 

However I think the current implementation is not well-specified. The "checkpoint version" is a sort of internal versioning scheme that's a combination of the epoch, pass, num edgepaths, and numPartitions. So if I set checkpoint_interval=10 and I have 1 pass, 1 edgepath, 1 partition, it will checkpoint every 10 epcohs. But if I have 20 passes then it will checkpoint twice per epoch.

What's a better way of specifying this that's not tied to the implementation? The easiest approach would be to only allow checkpointing at epoch intervals. It would also be nice if this code moved all the checkpoints to an epoch-labeled folder, e.g.  song_1.h5.10 --> checkpoints/epoch2/song_1.h5

**Comment by mingruimingrui** [comment]
@adamlerer You're right! The better implementation is to save current embedding states when epoch_idx is about to receive an increment. I've only considered the use case of 1 edge path, 1 edge chunk.

**Comment by mingruimingrui** [comment]
Now checkpoint files are copied into it's own separate folder. There however isn't any changes to the naming of the files, embedding files are still going to be called `
`embeddings_{entity}_{part}.v{version}.h5`

**Comment by mingruimingrui** [comment]
Fix, I was making sure that the code works and left a line of debug code in

File: docs/source/configuration_file.rst [hunk 1]
```diff
@@ -40,6 +40,10 @@ See :ref:`io-format` for more details.
 
   If set, it must be a path to a directory that contains initial values for the embeddings of all the entities of some types.
 
+- ``checkpoint_interval`` (type: string or null; default: ``null``)
```
Comment (lerks) [inline docs/source/configuration_file.rst]: You probably meant `integer`, not `string`.

File: torchbiggraph/train.py [hunk 2]
```diff
@@ -747,6 +747,18 @@ def log_status(msg, always=False):
...
         dlog("All parts of the checkpoint have been written")
 
+        if config.checkpoint_interval is not None:
+            # If it is the right interval for saving, we copy all current
+            # checkpoints into it's own separate folder
+            is_save_interval = epoch_idx % config.checkpoint_interval == 0
+            is_last_edge_path = \
+                (edge_path_idx + 1) == iteration_manager.num_edge_paths
+            is_last_edge_chunk = \
+                (edge_chunk_idx + 1) == iteration_manager.num_edge_chunks
```
Comment (lerks) [inline torchbiggraph/train.py]: It maybe clearer to add a `is_last_iteration_of_epoch` method (or a better name) to the `iteration_manager` and hide this code in there.

File: torchbiggraph/train.py [hunk 3]
```diff
@@ -747,6 +747,18 @@ def log_status(msg, always=False):
...
+
+            if is_save_interval & is_last_edge_path & is_last_edge_chunk:
```
Comment (lerks) [inline torchbiggraph/train.py]: You probably meant to use `and`, not `&`.

File: torchbiggraph/train.py [hunk 4]
```diff
@@ -747,6 +747,18 @@ def log_status(msg, always=False):
         sync.barrier()
         dlog("All parts of the checkpoint have been written")
 
+        if config.checkpoint_interval is not None:
+            # If it is the right interval for saving, we copy all current
+            # checkpoints into it's own separate folder
```
Comment (lerks) [inline torchbiggraph/train.py]: Nit: "its", not "it's".

Also, it's not "all current checkpoints" but "the entire current checkpoint".

File: docs/source/input_output.rst [hunk 5]
```diff
@@ -73,6 +73,8 @@ The latest complete checkpoint version is stored in an additional file in the sa
 
 Each checkpoint contains a JSON dump of the config that was used to produce it stored in the :file:`config.json` file.
 
+Old checkpoints will be automatically deleted unless specified through ``checkpoint_interval``
```
Comment (lerks) [inline docs/source/input_output.rst]: Could you make this clearer? Ideally, the doc should explain more than the `--help` text.

Like: "When a new checkpoint is saved, the previous one is deleted. If ``checkpoint_interval`` is specified, one checkpoint every ``checkpoint_interval`` epochs is preserved."

PS: `checkpoint_preservation_interval` is a longer but more self-descriptive name for that field.

**Comment by mingruimingrui** [comment]
@lerks Thanks for checking the code! As you suggest, old checkpoints will be moved instead of copied. A symlink is constructed in the old checkpoint's place to allow training to resume.

**Comment by mingruimingrui** [comment]
Don't merge this PR just yet, there's a bug where embeddings will be deleted if training is performed after it has already been completed. Here's what happens now.

Assuming `num_epochs = 5`, `torchbiggraph_train` would perform the training as per normal, move the final checkpoint to the `{checkpoint_path}/epoch_5` folder and create symlinks back to `{checkpoint_path}`.

When torchbiggraph_train performed again also at `num_epochs = 5`, the original code would do nothing after realizing that current checkpoint_version is 5. But now it would move the symlinks into `{checkpoint_path}/epoch_5`, effectively deleting the real embeddings.

**Comment by mingruimingrui** [comment]
I've tested these changes for a number of days, all bugs should now be ironed out!

**Comment by lerks** [comment]
Thanks for working on this! I'll be very busy in the next few days, I'm hoping I'll find some time to review this on Friday. Sorry to keep you waiting...

**Comment by mingruimingrui** [comment]
Hi sorry I was away for the last few days!

I interpret preserving every N epoch as saving the embedding states after N training cycles. With the current code structure, epoch_0 is the initialized states and epoch_1 is the state after 1 training cycle.

The reason for preserving epoch_0 is because when I first started using the library, I used the same directory for `init_path` and `checkpoint_path` and the initial embeddings got removed as a result. (My bad shouldn't have done that!) Currently, nothing would be done if `embeddings.v0.h5` is not found.

About point 3, epoch N is still saved correctly as `save_current_version` uses the argument `epoch_idx`. In practice, this means I'm preserving epoch N after N+1 is produced. I think I should rework this logic.

I've only considered the use case of the average Linux user. You're absolutely right on the part where we should avoid as much needless moving of files as possible.

I don't think that the behavior should be changed from preserving every N epoch to N-1, 2*N-1...

But let me list out the changes I'll definitely be making

- Change the name to `checkpoint_preservation_interval`.
- Place symlinks into `epoch_X` folders and keep original files in `checkpoint_path`
- Do preservation after `self.checkpoint_version` is increased

**Comment by mingruimingrui** [comment]
It came to my attention that if we want to create symlinks. windows users would have to provide 'admin permission'. It would make much more sense to do a restructuring of the way files are saved. So instead of saving `embeddings.h5` files to `checkpoint_path`, we save them directly to `checkpoint_path/epoch_x`

**Comment by lerks** [comment]
To be honest, we never tested that PBG works in Windows, and I don't think we'll be able to start doing it consistently. That's a way of saying that we don't officially support Windows. It may very well work and, in fact, nothing comes to mind that is unsupported on Windows. Did you try it or know someone who did? Does it work?

As for this pull request, I think we should focus on getting it to work properly under Linux, with an eye on avoiding it to break Windows even if that means a feature degradation. What I mean is that we should stick to the original idea of creating a symlink and, if that fails, print a warning but keep going. What do you think?

**Comment by mingruimingrui** [comment]
Sounds good to me! Let me work on this over the weekends

**Comment by mingruimingrui** [comment]
I think it would be wiser to preserving epoch N after epoch N+1 is produced. From a programming perspective, it makes little sense to perform this delayed operation but there are some merits which mainly concerns usage.

By preserving epoch N after N+1 is produced, we are effectively preserving a no longer required artifact. This makes it comparatively safer for the user to use the embeddings in the `epoch_N` folder. Take for instance a scenario where the user opened the `embeddings.h5` file from the `epoch_N` directory in `a` mode (the default mode using h5py). Other processes would no longer be able to access the file anymore. Usually, this wouldn't be an issue but if `torchbiggraph_train` still needs to read from those `embeddings.h5` files there would be problems!

Let me know what you think.

**Comment by lerks** [comment]
Sorry for the inexcusably long delay, I had missed that you had updated the PR. I've now finished polishing it and merged it, I hope it matches what you had set out to do.

**Pull request closed without merging**